BUG: reference count leaks in _ctypes

sunqm opened the pull request:
ctypes.POINTER function caches the C pointer types created by core._internal._ctypes class

https://github.com/python/cpython/blob/master/Modules/_ctypes/callproc.c#L1783

This leads to an extra reference count to the array object in function _get_void_ptr. Changing the call to ctypes.pointer to ctypes.byref should fix the problem in #13808

mattip commented:
Thanks for looking into this.
There is a failing test, and probably this change should get some new tests.
You might find it helpful to check your changes by running tests locally
```
python3 runtests.py
```

sunqm committed:
BUG: Fix _ctypes class cirular reference. (#13808)

In _ctypes class, ctypes.cast() was called twice. It causes circular reference
for _ctypes._data due to the CPython bug https://bugs.python.org/issue12836.

sunqm commented:
It does not fix the issue. PR #14469 fixes the circular reference issue
